docs: migrate multi-model deployment guide from gateway-api-inference-extension (#1604)
llm-d/llm-d: guides/multi-model/serving-multiple-inference-pools.md

# Serving Multiple Inference Pools

This guide assumes you completed the Getting Started guide before running the current guide.

> [!WARNING]
> **Unreleased/main branch**
>

> > This guide tracks **main** and is intended for users who want the very latest features and fixes and are comfortable with potential breakage.

Reviewer: Can we get rid of these? The guide should be stable.

If cutting a release is blocker then we should - though note that other guides also use nightlies.

A company may need to deploy multiple large language models (LLMs) in a cluster to support different workloads. For example, a Qwen model could power a chatbot interface, while a DeepSeek model might serve a recommendation application. Additionally, each base model may have multiple Low-Rank Adaptations ([LoRAs](https://www.ibm.com/think/topics/lora)). LoRAs associated with the same base model are served by the same backend inference server that hosts the base model. A LoRA name is also provided as the model name in the request body.

For serving multiple inference pools, the system needs to extract information such as the model name from the request body. This pattern of serving multiple models behind a single endpoint is common among providers and is generally expected by clients.

> For such model-aware routing, use the Body-Based Routing (BBR) component as described in this guide.

Reviewer 2: ```suggestion
For such model-aware routing, use the Inference-Payload-Processor (IPP) component as described in this guide.
```

## How

> The BBR extracts the model name from the request body, does a lookup of the base model in a configmap and adds this information in the `X-Gateway-Base-Model-Name` header. This header is then used for matching and routing the request to the appropriate `InferencePool` and its associated Endpoint Picker Extension (EPP) instances.

Reviewer 2: ```suggestion
The IPP extracts the model name from the request body, does a lookup of the base model in a configmap and adds this information in the `X-Gateway-Base-Model-Name` header. This header is then used for matching and routing the request to the appropriate `InferencePool` and its associated Endpoint Picker Extension (EPP) instances.
```

⚠️ **Note**: All model names, including base and LoRA names must be unique in order to be able to understand what is the correct `InferencePool` that should receive the request.

> ### Deploy Body-Based Routing Extension

Reviewer 2: ```suggestion
### Deploy Inference-Payload-Processor
```

   ```bash

>    export CHART_VERSION=v0

Reviewer 2: we don't have v0 (need to use released version):
```suggestion
   export CHART_VERSION=<SELECTED_VERSION>
```

   ```

<details>
<summary><b>GKE</b></summary>

```bash
export GATEWAY_PROVIDER=gke

> helm install body-based-router \

Reviewer 2: ```suggestion
helm install payload-processor \
```

--set provider.name=$GATEWAY_PROVIDER \
--version $CHART_VERSION \

> oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/body-based-routing

Reviewer 3: @shmuelk @nirrozenbaum I don't see dev helm charts or images created in https://github.com/llm-d/llm-d-inference-payload-processor; can we have those so we migrate from using the igw artifacts?

/cc @nilig

Reviewer 2: @ahg-g we migrated the helm charts https://github.com/llm-d/llm-d-inference-payload-processor/tree/main/config/charts/payload-processor., but:
- we didn't cut a release yet. hopefully once we do we should have at least a release version.
- we don't have (yet) automation to create helm chart tag from main.

Reviewer 3: @nirrozenbaum  do we have a release cut for 0.8.0?

Reviewer 2: > @nirrozenbaum do we have a release cut for 0.8.0?

we've just put in place the release automation few years ago.. 
I've created few days ago the first RC.

Reviewer 2: https://github.com/llm-d/llm-d-inference-payload-processor/pkgs/container/charts%2Fpayload-processor

I assume we should have the final release in few days

> oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/body-based-routing

Reviewer 2: ```suggestion
oci://ghcr.io/llm-d/charts/payload-processor
```

```

</details>

<details>
<summary><b>Istio</b></summary>

```bash
export GATEWAY_PROVIDER=istio

> helm install body-based-router \

Reviewer 2: ```suggestion
helm install payload-processor \
```

--set provider.name=$GATEWAY_PROVIDER \
--version $CHART_VERSION \

> oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/body-based-routing

Reviewer 2: ```suggestion
oci://ghcr.io/llm-d/charts/payload-processor
```

```

</details>

<details>
<summary><b>Agentgateway</b></summary>

> Agentgateway does not require the Body-Based Routing Extension, and instead natively implements Body-Based Routing.

Reviewer 2: ```suggestion
Agentgateway does not require the Inference-Payload-Processor Extension, and instead natively implements the same functionality.
```

> To use Body Based Routing, apply an `AgentgatewayPolicy`:

Reviewer 2: ```suggestion
To use AgentGateway, apply an `AgentgatewayPolicy`:
```

```bash
kubectl apply -f- <<EOF
apiVersion: agentgateway.dev/v1alpha1
kind: AgentgatewayPolicy
metadata:

>   name: bbr

Reviewer 2: ```suggestion
  name: ipp
```

spec:
  targetRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: inference-gateway
  traffic:
    phase: PreRouting
    transformation:
      request:
        set:
        - name: X-Gateway-Base-Model-Name
          value: |
            {"Qwen/Qwen3-32B": "Qwen/Qwen3-32B", "food-review-1": "Qwen/Qwen3-32B", "deepseek/DeepSeek-r1": "deepseek/DeepSeek-r1", "ski-resorts": "deepseek/DeepSeek-r1", "movie-critique": "deepseek/DeepSeek-r1"}[string(json(request.body).model)]
EOF
```

</details>

<details>
<summary><b>Other</b></summary>

```bash

> helm install body-based-router \

Reviewer 2: ```suggestion
helm install payload-processor \
```

--version $CHART_VERSION \

> oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/body-based-routing

Reviewer 2: ```suggestion
oci://ghcr.io/llm-d/charts/payload-processor
```

The example uses a vLLM simulator since this is the least common denominator configuration that can be run in every environment.
The manifest uses `deepseek/DeepSeek-r1` as base model with two LoRA adapters `ski-resorts` and `movie-critique`.

Deploy the second model server along with a mapping from LoRA adapters to the base model:

```bash
kubectl apply -f https://github.com/llm-d/llm-d/raw/main/guides/multi-model/manifests/sim-deployment.yaml
```

### Deploy the Second InferencePool and Endpoint Picker Extension

In order to create the `HttpRoute` mapping via the helm chart, one should use the experimental flag (`experimentalHttpRoute`) to specify the
base model that should be associated with the `InferencePool`.

Set the Helm chart version (unless already set).

   ```bash
   export IGW_CHART_VERSION=v0
   ```

<details>
<summary><b>GKE</b></summary>

```bash
export GATEWAY_PROVIDER=gke
helm install vllm-deepseek-r1 \
--dependency-update \
--set inferencePool.modelServers.matchLabels.app=vllm-deepseek-r1 \
--set provider.name=$GATEWAY_PROVIDER \
--set experimentalHttpRoute.enabled=true \
--set experimentalHttpRoute.baseModel=deepseek/DeepSeek-r1 \
--version $IGW_CHART_VERSION \

> oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/inferencepool

Reviewer 2: should be changed to llm-d chart

```

</details>

<details>
<summary><b>Istio</b></summary>

```bash
export GATEWAY_PROVIDER=istio
helm install vllm-deepseek-r1 \
--dependency-update \
--set inferencePool.modelServers.matchLabels.app=vllm-deepseek-r1 \
--set provider.name=$GATEWAY_PROVIDER \
--set experimentalHttpRoute.enabled=true \
--set experimentalHttpRoute.baseModel=deepseek/DeepSeek-r1 \
--version $IGW_CHART_VERSION \

> oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/inferencepool

Reviewer 2: ditto

```

</details>

<details>
<summary><b>Agentgateway</b></summary>

```bash
export GATEWAY_PROVIDER=none
helm install vllm-deepseek-r1 \
--dependency-update \
--set inferencePool.modelServers.matchLabels.app=vllm-deepseek-r1 \
--set provider.name=$GATEWAY_PROVIDER \
--set experimentalHttpRoute.enabled=true \
--set experimentalHttpRoute.baseModel=deepseek/DeepSeek-r1 \
--version $IGW_CHART_VERSION \

> oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/inferencepool

Reviewer 2: ditto

```

</details>

<details>
<summary><b>Other</b></summary>

```bash
helm install vllm-deepseek-r1 \
--dependency-update \
--set inferencePool.modelServers.matchLabels.app=vllm-deepseek-r1 \
--set experimentalHttpRoute.enabled=true \
--set experimentalHttpRoute.baseModel=deepseek/DeepSeek-r1 \
--version $IGW_CHART_VERSION \

> oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/inferencepool

Reviewer 2: ditto

```

</details>

After the installation, verify that you have two `InferencePools` and two EPP pods, one per base model type, running without errors

```bash
kubectl get inferencepools
```

```bash
kubectl get pods
```

### Upgrade the First InferencePool and Endpoint Picker Extension

Update the first model server mapping of the LoRA adapters to the base model:

```bash
kubectl apply -f https://github.com/llm-d/llm-d/raw/main/guides/multi-model/manifests/configmap.yaml
```

Run `helm upgrade` in order to update in place the HttpRoute mapping of the first `InferencePool`:

<details>
<summary><b>GKE</b></summary>

```bash
export GATEWAY_PROVIDER=gke

> helm upgrade vllm-qwen3-32b oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/inferencepool \

Reviewer 2: ditto

--dependency-update \
--set inferencePool.modelServers.matchLabels.app=vllm-qwen3-32b \
--set provider.name=$GATEWAY_PROVIDER \
--set experimentalHttpRoute.enabled=true \
--set experimentalHttpRoute.baseModel=Qwen/Qwen3-32B \
--version $IGW_CHART_VERSION
```

</details>

<details>
<summary><b>Istio</b></summary>

```bash
export GATEWAY_PROVIDER=istio

> helm upgrade vllm-qwen3-32b oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/inferencepool \

Reviewer 2: ditto

--dependency-update \
--set inferencePool.modelServers.matchLabels.app=vllm-qwen3-32b \
--set provider.name=$GATEWAY_PROVIDER \
--set experimentalHttpRoute.enabled=true \
--set experimentalHttpRoute.baseModel=Qwen/Qwen3-32B \
--version $IGW_CHART_VERSION
```

</details>

<details>
<summary><b>Agentgateway</b></summary>

```bash
export GATEWAY_PROVIDER=none

> helm upgrade vllm-qwen3-32b oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/inferencepool \

Reviewer 2: ditto

--dependency-update \
--set inferencePool.modelServers.matchLabels.app=vllm-qwen3-32b \
--set provider.name=$GATEWAY_PROVIDER \
--set experimentalHttpRoute.enabled=true \
--set experimentalHttpRoute.baseModel=Qwen/Qwen3-32B \
--version $IGW_CHART_VERSION
```

</details>

<details>
<summary><b>Other</b></summary>

```bash

> helm upgrade vllm-qwen3-32b oci://us-central1-docker.pkg.dev/k8s-staging-images/gateway-api-inference-extension/charts/inferencepool \

Reviewer 2: ditto
